### PR TITLE
[wasm] Addressing System.Runtime.Extensions failures

### LINKED ISSUE #38164: [WASM] Environment.UserName raises "not supported" exception
**Update:** UserName now fails with `System.IO.IOException : Not supported`

~~This is causing `System.Runtime.Extensions.Tests` to crash.~~

~~Crashes with:~~
<details>
```
at 205708 (dotnet.js:1858:32)
    at _emscripten_asm_const_iii (dotnet.js:1863:27)
    at wasm_logger (<anonymous>:wasm-function[7791]:0x1f3d0d)
    at eglib_log_adapter (<anonymous>:wasm-function[6303]:0x1abc50)
    at monoeg_g_logstr (<anonymous>:wasm-function[7225]:0x1ced8d)
    at monoeg_g_logv_nofree (<anonymous>:wasm-function[7223]:0x1cecb4)
    at monoeg_g_logv (<anonymous>:wasm-function[7222]:0x1cec11)
    at monoeg_g_log (<anonymous>:wasm-function[7226]:0x1ceded)
    at getpwuid_r (<anonymous>:wasm-function[1131]:0x6e9e5)
    at SystemNative_GetPwUidR (<anonymous>:wasm-function[7726]:0x1effbc)
```
</details>

~~I guess this API should throw PNSE in WASM?~~

Will disable the tests against this issue in the meantime. 

cc: @steveisok @lewing @marek-safar @akoeplinger

### PR DESCRIPTION
**Pull request by mdh1418**
Contributes towards #38422 
Fixes https://github.com/dotnet/runtime/issues/38164
This PR looks to address test failures in `System.Runtime.Extensions` when running on Browser wasm.
`Tests run: 6875, Errors: 0, Failures: 0, Skipped: 50. Time: 23.8534559s`
This PR's changes will be integrated with changes related to environment which would account for the one failure and some skips. @steveisok 

[Failures](https://github.com/dotnet/runtime/files/4897407/system_runtime_extensions_failures.txt)

### CODE REVIEW

File: src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs
Comment (jkotas) [inline src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs:235]: How is path normalization coupled with threading?

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (jkotas) [inline src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs:89]: Same here: How is globalization coupled with threading?

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs
Comment (jkotas) [inline src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs:12]: Should this rather throw PlatformNotSupportedException? It would be nice to have a tests that verifies the exception it throws.

File: src/libraries/System.Runtime.Extensions/tests/System/AppDomainTests.cs
Comment (jkotas) [inline src/libraries/System.Runtime.Extensions/tests/System/AppDomainTests.cs:125]: You can just fix this one by replacing `Assembly.GetExecutingAssembly().GetName().Name;` below with `DefaultDomain`.

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs
Comment (steveisok) [reply src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs:12]: It shouldn't.  In 3.2, we return "emscripten", which is what `gethostname` returns.  

I will have a follow up PR that addresses this.

File: src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs
Comment (jkotas) [inline src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs:235]: Nit: `{` should be on a new line

Comment (steveisok) [reply src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs:235]: I think this is more "if wasm".  If I understand it correctly, `Path.GetDirectoryName` is going to return "/" and going up a directory doesn't make a lot of sense.

Comment (jkotas) [reply src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs:235]: Agree. It should be `if (!PlatformDetection.IsBrowser)` then

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (steveisok) [reply src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs:89]: This needs some kind of ICU detection.  The problem is the tests are failing when invariant mode is enabled.  

@safern is working on disabling / enabling overall.  Perhaps we should remove the change from this PR for the time being.

File: src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs]: Thanks!

File: src/libraries/System.Runtime.Extensions/tests/System/AppDomainTests.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: Thanks!

File: src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs]: Done.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (steveisok) [inline src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs:89]: Can we wait to exclude?  It would be easy to miss when we add some sort of ICU feature detection.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (steveisok) [inline src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs:195]: Same re: ICU detection.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (steveisok) [inline src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs:207]: Same re: ICU detection.

File: src/libraries/tests.proj
Comment (steveisok) [inline src/libraries/tests.proj:83]: I would still exclude for now.  There are a few follow up items that will need to happen.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs]: Done.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs]: Done.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs]: Done.

File: src/libraries/tests.proj
Comment (mdh1418) [reply src/libraries/tests.proj]: Done.

File: src/libraries/System.Runtime.Extensions/tests/System/AppDomainTests.cs
Comment (jkotas) [inline src/libraries/System.Runtime.Extensions/tests/System/AppDomainTests.cs:47]: This seems to be using `RemoteExecutor`. Exclude it using `[ConditionalFact(typeof(RemoteExecutor)...`?

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs
Comment (jkotas) [reply src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs:12]: Is there issue to link to?

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.Exit.cs
Comment (jkotas) [inline src/libraries/System.Runtime.Extensions/tests/System/Environment.Exit.cs:44]: RemoteExecutor?

File: src/libraries/System.Runtime.Extensions/tests/System/AppDomainTests.cs
Comment (steveisok) [reply src/libraries/System.Runtime.Extensions/tests/System/AppDomainTests.cs:47]: Hmm... My guess is that we're going to need a different test even w/ RemoteExecutor enabled. Process makes sense when running on node, but in a browser, not so much.

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.Exit.cs
Comment (steveisok) [reply src/libraries/System.Runtime.Extensions/tests/System/Environment.Exit.cs:44]: Same as above.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (safern) [reply src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs]: Yeah, as port of: https://github.com/dotnet/runtime/pull/38976 I’m adding a PlatformDetection property to know if we’re running in Invariant so probably we should skip this test if that’s the case.

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs
Comment (steveisok) [reply src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs:12]: No, but I'll have a PR right after this one lands.

File: src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs
Comment (akoeplinger) [inline src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs:235]: nit: we could make this more generic by checking whether we're at the root:

```suggestion
                if (currentDirectory != Path.GetPathRoot(currentDirectory))
```

File: src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs
Comment (akoeplinger) [inline src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:207]: @steveisok this will also be reenabled as part of your PR right?

Comment (steveisok) [reply src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:207]: Yes.  Probably makes sense to just exclude this b/c I modified the test.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (safern) [reply src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs]: I merged my PR in case you want to use that property to skip these cases.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (akoeplinger) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:35]: same comment as above and doesn't need unsafe:

```suggestion
        // Matching what we returned for an earlier release
        public static string UserName => "web_user";
```

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:15]: Nit: Inconsistent formatting. All properties in this file can just use the single-line `=>`

Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:15]: I need to get up w/ the times ;-).

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs
Comment (steveisok) [reply src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs:12]: I just added my changes to this PR instead.  We felt it made more sense.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:13]: I'm not sure we want to expose this value

@lewing

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:19]: Can we return just empty string?

File: src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs]: Done

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:13]: I'd rather have it be "localhost" or better yet map, if we can, to location.hostname.  But this value is compatible to what's returned in 3.2

Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:13]: I'm not concern about this one, we are switching from Mono BCL to .NET Core anyway

File: src/libraries/System.Private.CoreLib/src/System/Environment.Unix.cs
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Unix.cs:27]: Does this work with Browser or does it always return same value?

File: src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs
Comment (marek-safar) [inline src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:224]: Assert.Equal instead?

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:19]: I think it makes more sense to stick w/ this or some kind of default.  I don't think it should be an empty string.

File: src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs]: Done

File: src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs
Comment (safern) [inline src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs:235]: Why do we need to do it this way, instead of just checking if we're not running in Browser?

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (safern) [inline src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs:90]: Actually can we change this to `PlatformDetection.IsNotInvariantGlobalization`? You will need to merge with master to get that. 

The reason for that is because this is also supported on Windows systems that don't have `ICU` and will end up running on NLS.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (safern) [inline src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs:195]: Ditto, use `PlatformDetection.IsNotInvariantGlobalization`

File: src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs
Comment (akoeplinger) [reply src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs:235]: it's cleaner as this isn't really about browser but rather what currentDirectory you're running the test in. E.g. if you ran this test in `C:\` on Windows it wouldn't work either.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (safern) [inline src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs:207]: Ditto.

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs
Comment (safern) [reply src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs:12]: Should we update this comment?

File: src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs
Comment (safern) [reply src/libraries/System.Runtime.Extensions/tests/System/IO/PathTests.cs:235]: I thought this was specific of Browser, since we didn't need this before.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs]: Done.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs]: Done.

File: src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/StringComparer.cs]: Done.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:13]: The Unix implementation already does:
```
Interop.Sys.GetHostName();
```
and SystemNative_GetHostName just calls:
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/Native/Unix/System.Native/pal_networking.c#L573
Why doesn't that work for wasm?

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs
Comment (mdh1418) [reply src/libraries/System.Runtime.Extensions/tests/System/Environment.MachineName.cs]: Updated to not skip the test, instead modified `GetComputerName` to return emscripten on Browser.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:13]: It does work, it just returns "emscripten".  We may be able to link something out by hard coding it into Environment.

**Review by akoeplinger — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:20]: It's not clear this is "some kind of default" is it?

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:13]: @stephentoub we cannot remove unused native code, calling pinvoke means every app will have the code included, in this case, to return a constant string which is not worth it.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:13]: Interop.Sys.GetHostName is used by Environment, by Dns, by PipeStream, by NetworkInformation, and by Process.  We're hardcoding all such call sites?  Or none of those are available on browser and this is the only one left?

Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:13]: I'll leave that to @steveisok to confirm but I think we turned all these assemblies into PNSE on browser

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:20]: No, it’s slightly better than emscripten 😜.  Open to suggestions.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (akoeplinger) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:13]: Correct, all of them except Dns are already PNSE and I'm working on doing that for Dns as well.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:20]: Suggestions: `Browser`, `.NET`, `UserName`

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:1826]: We could merge these together

File: src/libraries/System.Private.CoreLib/src/System/Environment.UnixOrBrowser.cs
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/Environment.UnixOrBrowser.cs:59]: Why is it in UnixOrBrowser? It seems to be needed on for Unix

**Review by marek-safar — CHANGES_REQUESTED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Environment.UnixOrBrowser.cs
Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System/Environment.UnixOrBrowser.cs:59]: Unless we're going to hard code it, it should be here.  Thoughts?

https://github.com/emscripten-core/emscripten/blob/master/system/lib/libc/musl/src/legacy/getpagesize.c#L5-L10

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (steveisok) [reply src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:1826]: Good catch

File: src/libraries/System.Private.CoreLib/src/System/Environment.UnixOrBrowser.cs
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Environment.UnixOrBrowser.cs:59]: This is about `string UserName`

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:13]: Ok, if we're going to hardcode it, I agree we should use something other than "emscripten".  Why match what GetHostName returns if we're trying not to use GetHostName, and if we're trying to hide the fact that it's implemented as unix under the covers with emscripten.  "localhost" sounds reasonable to me, or ".", or even empty string.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs:20]: Any of @marek-safar's suggestions sound reasonable to me.  Or maybe "Unknown".  I'm also fine with an empty string.

File: src/libraries/System.Private.CoreLib/src/System/Environment.UnixOrBrowser.cs
Comment (akoeplinger) [reply src/libraries/System.Private.CoreLib/src/System/Environment.UnixOrBrowser.cs:59]: @steveisok this should still be moved to `Environment.Unix.cs`.

**Review by akoeplinger — APPROVED** [review]

**Review by marek-safar — APPROVED** [review]

**Review by safern — APPROVED** [review]

**Pull request merged**
